neo/Ledger/BlockChain.State.cs

```diff
@@ -0,0 +1,181 @@
+using Akka.Actor;
+using Neo.IO;
+using Neo.Network.P2P;
+using Neo.Network.P2P.Payloads;
+using Neo.Persistence;
+using Neo.Persistence.LevelDB;
+using Neo.Trie.MPT;
+using System;
+using System.Collections.Generic;
+
+namespace Neo.Ledger
+{
+    public sealed partial class Blockchain : UntypedActor
+    {
+        public class ImportRoots { public IEnumerable<StateRoot> Roots; }
+        public uint StateHeight => currentSnapshot.StateHeight;
+        private uint StateRootEnableIndex => ProtocolSettings.Default.StateRootEnableIndex;
+        private readonly Dictionary<uint, StateRoot> stateRootCache = new Dictionary<uint, StateRoot>();
+
+        public StateRootState GetStateRoot(UInt256 block_hash)
+        {
+            var block = GetBlock(block_hash);
+            return block is null ? null : GetStateRoot(block.Index);
+        }
+
+        public StateRootState GetStateRoot(uint index)
+        {
+            return currentSnapshot.StateRoots.TryGet(index);
+        }
+
+        public bool GetStateProof(UInt256 root, StorageKey skey, out HashSet<byte[]> proof)
+        {
+            var trieReadOnlyDb = new TrieReadOnlyStore(Store, Prefixes.DATA_MPT);
+            var readOnlyTrie = new MPTReadOnlyTrie(root, trieReadOnlyDb);
+            return readOnlyTrie.GetProof(skey.ToArray(), out proof);
+        }
+
+        public bool VerifyProof(UInt256 root, byte[] key, HashSet<byte[]> proof, out byte[] value)
+        {
+            var result = MPTTrie.VerifyProof(root, key, proof, out value);
+            if (result) value = value.AsSerializable<StorageItem>().Value;
+            return result;
+        }
+
+        private RelayResultReason OnNewStateRoot(StateRoot state_root)
+        {
+            if (state_root.Index < StateRootEnableIndex || state_root.Index <= StateHeight) return RelayResultReason.Invalid;
+            if (state_root.Witness is null) return RelayResultReason.Invalid;
+            if (stateRootCache.ContainsKey(state_root.Index)) return RelayResultReason.AlreadyExists;
+            if (state_root.Index > Height || (state_root.Index > StateHeight + 1 && state_root.Index != StateRootEnableIndex))
+            {
+                stateRootCache.Add(state_root.Index, state_root);
```

Comment:
We should validate the signature of the state root before add it to the cache, otherwise one node could flood the network with invalid messages.